File: frontend/src/features/notifications/services/notificationService.ts
import { apiClient } from '../../../api/axios';

export interface NotificationDto {
  id: number;
  userId: number;
  workspaceId: number | null;
  notificationType: string;
  title: string;
  message: string;
  referenceType: string | null;
  referenceId: number | null;
  read: boolean;
  createdAt: string;
}

interface PaginatedResponse<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  number: number;
  size: number;
}

export const notificationService = {
  // GET /api/notifications — paginated list
  getNotifications: async (page = 0, size = 20, workspaceId?: number): Promise<PaginatedResponse<NotificationDto>> => {
    const params: Record<string, unknown> = { page, size };
    if (workspaceId) params.workspaceId = workspaceId;
    const response = await apiClient.get<PaginatedResponse<NotificationDto>>('/api/notifications', { params });
    return response.data;
  },

  // GET /api/notifications/unread
  getUnread: async (workspaceId?: number): Promise<NotificationDto[]> => {
    const params: Record<string, unknown> = {};
    if (workspaceId) params.workspaceId = workspaceId;
    const response = await apiClient.get<NotificationDto[]>('/api/notifications/unread', { params });
    return response.data;
  },

  // GET /api/notifications/unread/count
  getUnreadCount: async (workspaceId?: number): Promise<number> => {
    const params: Record<string, unknown> = {};
    if (workspaceId) params.workspaceId = workspaceId;
    const response = await apiClient.get<number>('/api/notifications/unread/count', { params });
    return response.data;
  },

  // PUT /api/notifications/{id}/read — mark single as read
  markAsRead: async (notificationId: number): Promise<void> => {
    await apiClient.put(`/api/notifications/${notificationId}/read`);
  },

  // PUT /api/notifications/read-all
  markAllAsRead: async (workspaceId?: number): Promise<void> => {
    const params: Record<string, unknown> = {};
    if (workspaceId) params.workspaceId = workspaceId;
    await apiClient.put('/api/notifications/read-all', null, { params });
  },

  // DELETE /api/notifications/{id}
  deleteNotification: async (notificationId: number): Promise<void> => {
    await apiClient.delete(`/api/notifications/${notificationId}`);
  },
};

File: backend/src/main/java/com/cloudstorage/backend/service/NotificationService.java
package com.cloudstorage.backend.service;

import com.cloudstorage.backend.dto.NotificationDto;
import com.cloudstorage.backend.entity.EntityType;
import com.cloudstorage.backend.entity.NotificationType;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface NotificationService {

    void createNotification(Long userId, Long workspaceId, NotificationType type, String title, String message, EntityType referenceType, Long referenceId);

    void notifyMention(Long userId, Long workspaceId, String title, String message, EntityType referenceType, Long referenceId);

    Page<NotificationDto> getUserNotifications(Long userId, Long workspaceId, Pageable pageable);

    List<NotificationDto> getUnreadNotifications(Long userId, Long workspaceId);

    void markAsRead(Long notificationId, Long userId);

    void markAllAsRead(Long userId, Long workspaceId);

    long getUnreadCount(Long userId, Long workspaceId);

    void deleteNotification(Long notificationId, Long userId);
}
